### PR TITLE
Add Metadata type for holding annotations and renames

### PR DESCRIPTION
**Pull request by jackkoenig**
Move those fields from CircuitState to Metadata

Here's Metadata, PR-ed against the refactor annotations branch

Some things get kind of annoying to express, like adding annotations to the ones already there:
```scala
val newAnnos = ...
// Before
state.copy(annotations = state.annotations ++ newAnnos)
// After
state.copy(metadata = state.metadata.copy(annotations = state.metadata.annotations ++ newAnnos))
```

Thoughts?

### CODE REVIEW

File: src/main/scala/firrtl/Compiler.scala [hunk 1]
```diff
@@ -105,20 +105,28 @@ object AnnotationSeq {
...
+/** Metadata associated with a Firrtl circuit
+  *
+  * @param annotations The current collection of [[firrtl.annotations.Annotation Annotation]]
+  * @param renames A map of [[firrtl.annotations.Named Named]] things that have been renamed.
+  *        Generally only a return value from [[Transform]]s
+  */
+case class Metadata(annotations: Seq[Annotation], renames: Option[RenameMap])
+object Metadata {
+  def apply(annotations: Seq[Annotation]) = new Metadata(annotations, None)
+}
```
Comment (azidar) [inline src/main/scala/firrtl/Compiler.scala]: Can you add a 
```scala
def add(newAnnos: Seq[Annotation]): Metadata = apply(annotations ++ newAnnos)
```

File: src/main/scala/firrtl/Compiler.scala [hunk 2]
```diff
@@ -105,20 +105,28 @@ object AnnotationSeq {
...
-  * @param renames A map of [[firrtl.annotations.Named Named]] things that have been renamed.
-  *   Generally only a return value from [[Transform]]s
+  * @param metadata Holds auxiliary information related to the circuit
   */
 case class CircuitState(
     circuit: Circuit,
     form: CircuitForm,
-    annotations: AnnotationSeq,
-    renames: Option[RenameMap]) {
+    metadata: Metadata) {
```
Comment (azidar) [inline src/main/scala/firrtl/Compiler.scala]: What do you think about pulling form into Metadata? I think it makes sense, as its actually metadata on the circuit.

File: src/main/scala/firrtl/passes/Uniquify.scala [hunk 3]
```diff
@@ -370,7 +370,7 @@ object Uniquify extends Transform {
 
     sinfo = c.info
     val result = Circuit(c.info, c.modules map uniquifyPorts(renames) map uniquifyModule(renames), c.main)
-    CircuitState(result, outputForm, state.annotations, Some(renames))
+    CircuitState(result, outputForm, state.metadata.copy(renames = Some(renames)))
```
Comment (azidar) [inline src/main/scala/firrtl/passes/Uniquify.scala]: I'd add another `def add(newRenames)` to simplify this syntax.

File: src/main/scala/firrtl/transforms/DeadCodeElimination.scala [hunk 4]
```diff
@@ -316,17 +316,17 @@ class DeadCodeElimination extends Transform {
     // Preserve original module order
     val newCircuit = c.copy(modules = c.modules.flatMap(m => modulesxMap.get(m.name)))
 
-    state.copy(circuit = newCircuit, renames = Some(renames))
+    state.copy(circuit = newCircuit, metadata = state.metadata.copy(renames = Some(renames)))
```
Comment (azidar) [inline src/main/scala/firrtl/transforms/DeadCodeElimination.scala]: we could add `def annotations: Seq[Annotation] = metadata.annotations` to CircuitState, which would simplify this syntax as well.

**Comment by jackkoenig** [comment]
We decided against this approach, instead we're going to make some changes to CircuitState.

**Pull request closed without merging**